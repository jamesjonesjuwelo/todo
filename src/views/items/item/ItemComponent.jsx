import React, { Component } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import styled from 'styled-components';
import { updateItem, deleteItem } from '../../../store/action/notesAction';
import bindActionCreators from '../../../../util/bindActionCreators';
import IconComponent from '../../../component/icon/IconComponent';
import InputComponent from '../../../component/input/ItemInputComponent';

export default class ItemComponent extends Component {

  state = {
    descriptionInputVisible: false,
    editIconVisible: false,
    description: ''
  };

  componentDidMount() {
    const { item: { description } } = this.props;
    this.setState({ description });
  }

  toggleDescriptionInput = () => {
    this.setState({
      descriptionInputVisible: !this.state.descriptionInputVisible
    });

  };

  toggleEditIcon = boolean => {
    if (!this.state.descriptionInputVisible)
      this.setState({ editIconVisible: boolean})
  };

  handleChange = value =>
    this.setState({ description: value });


  updateItem = () => {
    const { item, updateItem } = this.props;
    const updateInfo = {
      item,
      description: this.state.description
    };
    updateItem(updateInfo);
    this.toggleDescriptionInput();
  };

  deleteItem = () => {
    const { item, deleteItem } = this.props;
    deleteItem(item);
  };
  
  completeItem = () => {
    const { item, updateItem } = this.props;
    const updateInfo = {
      item,
      completed: true
    };
    updateItem(updateInfo);
  };

  render() {
    const { item: { name, description, createdAt, completed } } = this.props,
      { descriptionInputVisible, editIconVisible } = this.state,
      opacity = completed ? 0.3 : 1,
      tick = 'tick',
      edit = 'edit',
      cross = 'cross';

    return (
      <Container
        onMouseEnter={() => this.toggleEditIcon(true)}
        onMouseLeave={() => this.toggleEditIcon(false)}
        style={{ opacity }}
      >
        {editIconVisible && !completed
        ? <IconComponent
          method={this.completeItem}
          src={`/icons/${tick}.png`}
          styles={completeTaskTickIconStyles}
        />
        : <span />}
        <TextContainer>
          <CreatedAt>{createdAt}</CreatedAt>
          <DescriptionContainer>
            {!descriptionInputVisible
              ? <Description>{description}</Description>
              : <InputComponent
                  placeholder={this.state.description}
                  value={this.state.description}
                  handleChange={val => this.handleChange(val)}
                  triggerMethod={() => this.updateItem()}
                />}
            {editIconVisible && !completed
             ? <IconComponent
                  method={this.toggleDescriptionInput}
                  src={`/icons/${edit}.png`}
                  styles={editIconStyles}
              />
             : <span />
            }
            {descriptionInputVisible
              ? (<IconComponent
                  method={this.updateItem}
                  src={`/icons/${tick}.png`}
                  styles={tickStyles}
              />)
              : <span />}
          </DescriptionContainer>
          <Name>{name}</Name>
          {editIconVisible
            ? <IconComponent
              method={() => this.deleteItem()}
              src={`/icons/${cross}.png`}
              styles={closeIconStyles}
            />
            : <span />}
        </TextContainer>
      </Container>

    )
  }
}

const mapStateToProps = (state) => ({});

const mapDispatchToProps = (dispatch, props) => bindActionCreators({
  updateItem,
  deleteItem
}, dispatch, props);

const enhance = compose(
  connect(mapStateToProps, mapDispatchToProps)
);

const faintOrange = 'rgba(231, 111, 81, 0.3)';

const transition = `
    -moz-transition: all 0.1s ease-in;
    -o-transition: all 0.1s ease-in;
    -webkit-transition: all 0.1s ease-in;
    transition: all 0.1s ease-in;
`;

const Container = styled.div`
  height: 100px;
  padding: 10px;
  position: relative;
  border-bottom: 1px solid ${faintOrange};
  
  :hover {
    background-color: ${faintOrange};
  }
  
  ${transition}
`;

const TextContainer = styled.div`
  width: 77.5%;
  padding-left: 2.5%;
`;

const zeroMargin = 'margin: 0';

const DescriptionContainer = styled.div`
  display: flex;
  flex-direction: row;
`;

const Description = styled.p`
  ${zeroMargin}
  font-weight: bold;
`;

const Name = styled.p`
font-size: 12px;
  
`;

const CreatedAt = styled.p`
  font-size: 10px;
  color: grey;
`;

const editIconStyles = `
  opacity: 0.5;
  margin-top: 3px;
  width: 10px;
  height: 10px;
  
  ${transition}
`;

const tickStyles = `
  opacity: 1;
  width: 15px;
  height: 15px;
`;

const completeTaskTickIconStyles = `
  ${tickStyles}
  position: absolute;
  top: 50px;
  left: 2px;
`;

const closeIconStyles = `
  opacity: 0.3;
  width: 15px;
  height: 15px;
  position: absolute;
  right: 10px;
  top: 10px;
  
  ${transition}
`;

export const ItemComponentRoute = enhance(ItemComponent);


