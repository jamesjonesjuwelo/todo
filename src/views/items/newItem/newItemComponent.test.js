import React from 'react';
import { shallow } from 'enzyme';
import ItemComponent from './NewItemComponent';
import 'jest-styled-components';

const props = {
  createItem: jest.fn()
};

describe('New Item Component', () => {
  let wrapper;
  
  beforeEach(() => {
    wrapper = shallow( <ItemComponent {...props} /> );
  });
  
  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  
  test('handling name change', () => {
    wrapper.instance().handleChange({ type: 'name', val: 'Test one' });
    expect(wrapper.state().name).toEqual('Test one');
  });
  
  test('handling description change', () => {
    wrapper.instance().handleChange({ type: 'description', val: 'Test two' });
    expect(wrapper.state().description).toEqual('Test two');
  });
  
  test('create new item with zero args', () => {
    wrapper.instance().createNewItem();
    expect(wrapper.state().error).toEqual(true);
  });
});