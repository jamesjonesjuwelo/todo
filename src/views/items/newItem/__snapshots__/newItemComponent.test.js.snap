// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Item Component renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NewItemComponent
    createItem={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <styled.div>
          <Input
            handleChange={[Function]}
            placeholder="What to do?"
            styling={
              Object {
                "borderBottomLeftRadius": "30px",
                "borderTopLeftRadius": "30px",
                "width": "60%",
              }
            }
            triggerMethod={[Function]}
            value=""
          />
          <Input
            handleChange={[Function]}
            placeholder="Name"
            styling={
              Object {
                "borderRight": "1px solid white",
                "width": "20%",
              }
            }
            triggerMethod={[Function]}
            value=""
          />
          <AddButtonComponent
            createNewItem={[Function]}
          />
        </styled.div>,
        <span />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            Array [
              <Input
                handleChange={[Function]}
                placeholder="What to do?"
                styling={
                  Object {
                    "borderBottomLeftRadius": "30px",
                    "borderTopLeftRadius": "30px",
                    "width": "60%",
                  }
                }
                triggerMethod={[Function]}
                value=""
              />,
              <Input
                handleChange={[Function]}
                placeholder="Name"
                styling={
                  Object {
                    "borderRight": "1px solid white",
                    "width": "20%",
                  }
                }
                triggerMethod={[Function]}
                value=""
              />,
            ],
            <AddButtonComponent
              createNewItem={[Function]}
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "input_description",
            "nodeType": "function",
            "props": Object {
              "handleChange": [Function],
              "placeholder": "What to do?",
              "styling": Object {
                "borderBottomLeftRadius": "30px",
                "borderTopLeftRadius": "30px",
                "width": "60%",
              },
              "triggerMethod": [Function],
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "input_name",
            "nodeType": "function",
            "props": Object {
              "handleChange": [Function],
              "placeholder": "Name",
              "styling": Object {
                "borderRight": "1px solid white",
                "width": "20%",
              },
              "triggerMethod": [Function],
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "createNewItem": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "span",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <styled.div>
            <Input
              handleChange={[Function]}
              placeholder="What to do?"
              styling={
                Object {
                  "borderBottomLeftRadius": "30px",
                  "borderTopLeftRadius": "30px",
                  "width": "60%",
                }
              }
              triggerMethod={[Function]}
              value=""
            />
            <Input
              handleChange={[Function]}
              placeholder="Name"
              styling={
                Object {
                  "borderRight": "1px solid white",
                  "width": "20%",
                }
              }
              triggerMethod={[Function]}
              value=""
            />
            <AddButtonComponent
              createNewItem={[Function]}
            />
          </styled.div>,
          <span />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              Array [
                <Input
                  handleChange={[Function]}
                  placeholder="What to do?"
                  styling={
                    Object {
                      "borderBottomLeftRadius": "30px",
                      "borderTopLeftRadius": "30px",
                      "width": "60%",
                    }
                  }
                  triggerMethod={[Function]}
                  value=""
                />,
                <Input
                  handleChange={[Function]}
                  placeholder="Name"
                  styling={
                    Object {
                      "borderRight": "1px solid white",
                      "width": "20%",
                    }
                  }
                  triggerMethod={[Function]}
                  value=""
                />,
              ],
              <AddButtonComponent
                createNewItem={[Function]}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "input_description",
              "nodeType": "function",
              "props": Object {
                "handleChange": [Function],
                "placeholder": "What to do?",
                "styling": Object {
                  "borderBottomLeftRadius": "30px",
                  "borderTopLeftRadius": "30px",
                  "width": "60%",
                },
                "triggerMethod": [Function],
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "input_name",
              "nodeType": "function",
              "props": Object {
                "handleChange": [Function],
                "placeholder": "Name",
                "styling": Object {
                  "borderRight": "1px solid white",
                  "width": "20%",
                },
                "triggerMethod": [Function],
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "createNewItem": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "span",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
